google/cloud/aiplatform/models.py

```diff
@@ -15,9 +15,213 @@
+                Should be a dictionary where keys are environment variable names
+                and values are environment variable values for those names.
+            serving_container_ports: Optional[Sequence[int]]=None,
+                Declaration of ports that are exposed by the container. This field is
+                primarily informational, it gives AI Platform information about the
+                network connections the container uses. Listing or not a port here has
+                no impact on whether the port is actually exposed, any port listening on
+                the default "0.0.0.0" address inside a container will be accessible from
+                the network.
+            project: Optional[str]=None,
+                Project to upload this model to. Overrides project set in
+                aiplatform.init.
+            location: Optional[str]=None,
+                Location to upload this model to. Overrides location set in
+                aiplatform.init.
+            credentials: Optional[auth_credentials.Credentials]=None,
+                Custom credentials to use to upload this model. Overrides credentials
+                set in aiplatform.init.
+        Returns:
+            model: Instantiated representation of the uplaoded model resource.
+        """
+
+        api_client = cls._instantiate_client(location, credentials)
+        env = None
+        ports = None
+
+        if serving_container_environment_variables:
+            env = [
+                env_var.EnvVar(name=str(key), value=str(value))
+                for key, value in serving_container_environment_variables.items()
+            ]
+        if serving_container_ports:
+            ports = [
+                gca_model.Port(container_port=port) for port in serving_container_ports
+            ]
+
+        container_spec = gca_model.ModelContainerSpec(
+            image_uri=serving_container_image_uri,
+            command=serving_container_command,
+            args=serving_container_args,
+            env=env,
+            ports=ports,
+            predict_route=serving_container_predict_route,
+            health_route=serving_container_health_route,
+        )
+
+        managed_model = gca_model.Model(
+            display_name=display_name,
+            description=description,
+            artifact_uri=artifact_uri,
+            container_spec=container_spec,
+        )
+
+        lro = api_client.upload_model(
+            parent=initializer.global_config.get_resource_parent(project, location),
+            model=managed_model,
+        )
+
+        managed_model = lro.result()
+        fields = utils.extract_fields_from_resource_name(managed_model.model)
+        return cls(
+            model_name=fields.id, project=fields.project, location=fields.location
+        )
+
+    # TODO(b/169782716) add support for deployment when Endpoint class complete
+    def deploy(self):
```

Comment:
this is one place where the resource -> client mapping fails, since we need an EndpointServiceClient to deploy a model.

Reply:
That client is not mapped to the Model resource because this method returns an Endpoint class with EndpointServiceClient. That Endpoint class will either be passed in or instantiated.